### PR TITLE
Michelle's refactoring suggestions

### PR DESCRIPTION
**Pull request by mweidling**
In this PR the following changes have been made:

- moved config to a separate TOML file for easier customization (0f92cf9e6788e1ab8c3975ca5027d5a47af27e0d)
- added an .editorconfig file for better collaboration (3fd01de32899c99615ab93885b485eb3eb5a9612)
- sorted the import statements because my linter was complaining (b3da197c36d7ffcd44fa51ecbb44035ceff8b6f7)
- made a big round of changing indentation style from tabs to spaces (as it is recommended by PEP8) and all the other things that my linter complained about (21df5524e9b39f082316b2883fb1fa6547b5216c)
- added error codes to the validation because when I worked with the converter and had validation errors it was sometimes hard to find the place where something went wrong. error codes might make that easier (7a60c6f962c0abba8bd793779f92e73ed1677bc4)

Feel free to cherry-pick whatever seems right for you.

### CODE REVIEW

**Comment by MehmedGIT** [comment]
Hi Michelle! Thanks for your contribution. I went over the changes you suggested and found out a few things that could be further improved:

1. In the `config.toml`, we cannot simply replace `$HOME` and `$projectDir` with `.`. The project directory from the viewpoint of Nextflow may not be always inside the home or current directory. Moreover, each Nextflow process gets its own separate work environment, independent of other processes. Thus, we need to provide proper environment variables and let Nextflow handle them afterward -> `$projectDir`. The same argument is valid for `$HOME`. Even if you replace the home environment variable with `~`, Nextflow will complain that it cannot evaluate `~` to a path. I guess this is due to the implementation details of Nextflow. 

I like the idea of having a separate configuration file to set the proper paths. However, we must clearly specify why `$HOME` and `$projectDir` have to stay. Moreover, it may be confusing for the user why we need an escape character before `$HOME` but not before `$projectDir`. The answer is provided here: https://www.nextflow.io/docs/latest/process.html#script. 

```Since Nextflow uses the same Bash syntax for variable substitutions in strings, you must manage them carefully depending on whether you want to evaluate a Nextflow variable or a Bash variable.```

2. The token error numbers should ideally match with the number of the specific rule that was broken - rule numbers given in the comments above the validator methods. We may also extend the rules or provide them in a separate file to keep things clean.

3. I still do not understand how the imports are sorted. Seems they do not follow alphabetical order. Any ideas according to what the Linter sorted them? 

Everything else seems fine. Please implement the suggested changes and I will merge your PR.

**Comment by mweidling** [comment]
> Hi Michelle! Thanks for your contribution. I went over the changes you suggested and found out a few things that could be further improved:
> 
>     1. In the `config.toml`, we cannot simply replace `$HOME` and `$projectDir` with `.`. The project directory from the viewpoint of Nextflow may not be always inside the home or current directory. Moreover, each Nextflow process gets its own separate work environment, independent of other processes. Thus, we need to provide proper environment variables and let Nextflow handle them afterward -> `$projectDir`. The same argument is valid for `$HOME`. Even if you replace the home environment variable with `~`, Nextflow will complain that it cannot evaluate `~` to a path. I guess this is due to the implementation details of Nextflow.
> 
> 
> I like the idea of having a separate configuration file to set the proper paths. However, we must clearly specify why `$HOME` and `$projectDir` have to stay. Moreover, it may be confusing for the user why we need an escape character before `$HOME` but not before `$projectDir`. The answer is provided here: https://www.nextflow.io/docs/latest/process.html#script.
> 
> `Since Nextflow uses the same Bash syntax for variable substitutions in strings, you must manage them carefully depending on whether you want to evaluate a Nextflow variable or a Bash variable.`
> 
>     2. The token error numbers should ideally match with the number of the specific rule that was broken - rule numbers given in the comments above the validator methods. We may also extend the rules or provide them in a separate file to keep things clean.
> 
>     3. I still do not understand how the imports are sorted. Seems they do not follow alphabetical order. Any ideas according to what the Linter sorted them?
> 
> 
> Everything else seems fine. Please implement the suggested changes and I will merge your PR.

ad 1) Sure, I can re-introduce the original settings to the TOML.

ad 2) I will adjust the numbering according to the rule numbers. I will also think about a better solution for the rules but this shouldn't impede this PR IMO.

ad 3) I do not understand them as well. :smile: I use Pylint for VS Code and its "Sort import" functionality. [According to the docs](https://code.visualstudio.com/docs/python/editing#_sort-imports) it should sort them alphabetically, but well.

File: config.toml [hunk 1]
```diff
@@ -1,3 +1,6 @@
-venv_path = "./venv/bin/activate"
-workspace_path = "./ocrd-workspace/"
-mets_path = "./ocrd-workspace/mets.xml"
\ No newline at end of file
+# See https://www.nextflow.io/docs/latest/process.html#script for the difference
+# between NextFlow and Bash variables.
+venv_path = "\\$HOME/venv37-ocrd/bin/activate"
```
Comment (MehmedGIT) [inline config.toml:3]: A single escape character is needed before $HOME, not two.

File: ocrd_validator.py [hunk 2]
```diff
@@ -124,7 +124,7 @@ def __validate_single_len_token_symbols(self, line_index, token_index, token):
         if token in (QM, BACKSLASH):
             return True
         else:
-            info = f"TOKEN_ERROR01: Invalid token: {token}"
+            info = f"TOKEN_ERROR_RULE_01: Invalid token: {token}"
```
Comment (MehmedGIT) [inline ocrd_validator.py:127]: I think we need a distinction between symbol validation rules and just validation rules. I would change "TOKEN_ERROR_RULE_01" to "TOKEN_SYMBOL_ERROR_RULE_01"

File: ocrd_validator.py [hunk 3]
```diff
@@ -134,7 +134,7 @@ def __validate_double_len_token_symbols(self, line_index, token_index, token):
             if token[1] in ['I', 'O', 'P']:
                 return True
             else:
-                info = f"TOKEN_ERROR02: Invalid token: {token}"
+                info = f"TOKEN_ERROR_RULE_02: Invalid token: {token}"
```
Comment (MehmedGIT) [inline ocrd_validator.py:137]: The same here: "TOKEN_SYMBOL_ERROR_RULE_02".

File: ocrd_validator.py [hunk 4]
```diff
@@ -144,7 +144,7 @@ def __validate_double_len_token_symbols(self, line_index, token_index, token):
     def __validate_token_symbols(self, line_index, token):
         for char in token:
             if char not in VALID_CHARS:
-                info = f"TOKEN_ERROR03: Invalid token: {token}"
+                info = f"TOKEN_ERROR_RULE_03: Invalid token: {token}"
```
Comment (MehmedGIT) [inline ocrd_validator.py:147]: Here too.

File: ocrd_validator.py [hunk 5]
```diff
@@ -205,33 +205,33 @@ def __validate_min_amount_of_tokens(self, line_num, line, min_amount):
         if len(line) < min_amount:
             expected = f"At least {min_amount} tokens."
             got = f"{len(line)} tokens"
-            hint = 'TOKEN_ERROR04: Each line must have a processor call, an input file, and an output file'
+            hint = 'LINE_ERROR_RULE_02: Each line must have a processor call, an input file, and an output file'
```
Comment (MehmedGIT) [inline ocrd_validator.py:208]: The hint section should not include the error rule. Better to create an info variable and pass that variable to the print_syntax_error method. E.g., `info = "LINE_ERROR_RULE_02: Not enough amount of tokens"`

The hint should remain in this format:
`hint = "Each line must ..."`

And, btw, maybe we do not have to specify TOKEN/LINE in front of the "ERROR_RULE" to have a similar standard. This is just an idea, for now, you can keep it as you wish.

File: ocrd_validator.py [hunk 6]
```diff
@@ -205,33 +205,33 @@ def __validate_min_amount_of_tokens(self, line_num, line, min_amount):
...
             self.__print_syntax_error(line_num, expected=expected, got=got, hint=hint)
             sys.exit(2)
 
     def __validate_QM_tokens(self, line_num, first_qm_token, last_qm_token):
         if not first_qm_token == QM:
-            hint = 'TOKEN_ERROR05: OCR-D commands must start with a quotation mark'
+            hint = 'TLINE_ERROR_RULE_03: OCR-D commands must start with a quotation mark'
             self.__print_syntax_error(line_num, expected=QM, got=first_qm_token, hint=hint)
```
Comment (MehmedGIT) [inline ocrd_validator.py:215]: The same as above. A separate info variable should be created and passed in the method. The info should report the error and what caused the error. The hint section provides more details on how to fix it.
E.g., `info = "LINE_ERROR_RULE_03: Wrong start of the OCR-D command"`

File: ocrd_validator.py [hunk 7]
```diff
@@ -205,33 +205,33 @@ def __validate_min_amount_of_tokens(self, line_num, line, min_amount):
...
             sys.exit(2)
         if not last_qm_token == QM:
-            hint = 'TOKEN_ERROR06: OCR-D commands must end with a quotation mark'
+            hint = 'LINE_ERROR_RULE_11: OCR-D commands must end with a quotation mark'
```
Comment (MehmedGIT) [inline ocrd_validator.py:218]: The same here. The only difference in the info will be in "Wrong end of the OCR-D command"

File: ocrd_validator.py [hunk 8]
```diff
@@ -205,33 +205,33 @@ def __validate_min_amount_of_tokens(self, line_num, line, min_amount):
...
             self.__print_syntax_error(line_num, expected=QM, got=last_qm_token, hint=hint)
             sys.exit(2)
 
     def __validate_BACKSLASH_token(self, line_num, token):
         if not token == BACKSLASH:
-            hint = 'TOKEN_ERROR07: OCR-D lines must end with a backslash, except the last line'
+            hint = 'LINE_ERROR_RULE_12: OCR-D lines must end with a backslash, except the last line'
```
Comment (MehmedGIT) [inline ocrd_validator.py:224]: The same here. "Wrong end of OCR-D line"

File: ocrd_validator.py [hunk 9]
```diff
@@ -205,33 +205,33 @@ def __validate_min_amount_of_tokens(self, line_num, line, min_amount):
...
             self.__print_syntax_error(line_num, expected=BACKSLASH, got=token, hint=hint)
             sys.exit(2)
 
     def __validate_ocrd_processor_token(self, line_num, token):
         ocrd_processor = f"ocrd-{token}"
         if ocrd_processor not in OCRD_PROCESSORS:
-            hint = 'TOKEN_ERROR08: Ocrd-process is spelled incorrectly or does not exists'
+            hint = 'LINE_ERROR_RULE_01: `ocrd process` is spelled incorrectly or does not exists'
```
Comment (MehmedGIT) [inline ocrd_validator.py:231]: The same.

File: ocrd_validator.py [hunk 10]
```diff
@@ -269,7 +269,7 @@ def __validate_iop_tokens(self, line_num, line):
                 processed_tokens.append(next_token)
                 processed_tokens.append(next_token2)
             else:
-                info = "TOKEN_ERROR10: Unknown token found: {current_token}"
+                info = "INPUT_OUTPUT_ERROR: Unknown token found: {current_token}"
```
Comment (MehmedGIT) [inline ocrd_validator.py:272]: This is rather "UNEXPECTED_TOKEN_ERROR"

File: ocrd_validator.py [hunk 11]
```diff
@@ -286,14 +286,14 @@ def __validate_parameter_token(self, line_num, token, next_token, next_token2):
     def __is_valid_follow_up_token(self, line_num, previous_token, token, token2=None):
         # These tokens cannot be arguments passed to the previous_token
         forbidden_follow_ups = [QM, BACKSLASH, '-I', '-O', '-P']
-        info = f"TOKEN_ERROR11: Wrong or missing arguments to {previous_token}"
+        info = f"FOLLOW_UP_TOKEN_ERROR: Wrong or missing arguments to {previous_token}"
```
Comment (MehmedGIT) [inline ocrd_validator.py:289]: "ARGUMENT_TOKEN_ERROR"

File: ocrd_validator.py [hunk 12]
```diff
@@ -306,12 +306,13 @@ def __is_valid_follow_up_token(self, line_num, previous_token, token, token2=Non
                 sys.exit(2)
 
         return True
- 
+
     def __print_io_duplicate_error(self, line_num, index, token):
-        info = f"TOKEN_ERROR15: Duplicate {token} found!"
+        info = f"DUPLICATE_TOKENS_ERROR: Duplicate {token} found!"
```
Comment (MehmedGIT) [inline ocrd_validator.py:311]: "DUPLICATE_TOKEN_ERROR" just to try keeping the standard.

**Review by MehmedGIT — COMMENTED** [review]
Almost there :)

File: config.toml (hunk 1 above)
Comment (mweidling) [reply config.toml:3]: In TOML `\$` is not an allowed control sequence.

If you load the config file with `tomli`, it resolves correctly:

```python
import tomli

with open('config.toml', mode='rb') as fp:
    config = tomli.load(fp)

print(config['venv_path']
```

This yields `\$HOME/venv37-ocrd/bin/activate`. That should be correct, right?

**Pull request merged**